### PR TITLE
Add missing type hints to `synapse.appservice`

### PR DESCRIPTION
**Pull request by clokep**
This got a bit out of hand since there were some untyped dicts being tossed around. I introduced some attrs classes to clean this up.

This should be reviewable commit-by-commit.

### CODE REVIEW

File: synapse/appservice/__init__.py
Comment (clokep) [inline synapse/appservice/__init__.py:49]: This class doesn't seem to add much, I wonder if we should just inline these onto the `ApplicationService` class.

Comment (anoadragon453) [reply synapse/appservice/__init__.py:49]: As a data point, I could have used it when writing tests at https://github.com/matrix-org/synapse/pull/11215/files#diff-0b8c043365754e58302f0701c4e2f273dc190f516a326b42496b68a09fe26ef6R555 :slightly_smiling_face: 

But an `Dict[ApplicationService.NS_LIST, Namespace]` isn't too bad to write either.

Comment (anoadragon453) [reply synapse/appservice/__init__.py:49]: > Dict[ApplicationService.NS_LIST, Namespace]

Oh, I assumed this would restrict the entries to just the three types of allowed namespace strings, rather than allowing any str. But perhaps that's not possible.

Comment (clokep) [reply synapse/appservice/__init__.py:49]: You need to use an enum to do that -- that could be a solution though.

Comment (anoadragon453) [reply synapse/appservice/__init__.py:49]: Oh right, yes indeed! That could be useful to have available.

**Comment by clokep** [comment]
Going to hold off on this until #11215 is merged!

File: synapse/appservice/__init__.py
Comment (DMRobertson) [reply synapse/appservice/__init__.py:49]: > You need to use an enum to do that -- that could be a solution though.

Or a `Literal["name1", "name2", "name3"]`.

I was really keen to use `StrEnum` but Sean found that it didn't work well with interning. See #11356.

**Comment by clokep** [comment]
Discussed this with @anoadragon453 and he suggested we land this without waiting for #11215 since that's dragging a bit.

@DMRobertson could you take another look?

**Review by DMRobertson — COMMENTED** [review]
A few minor things:

File: synapse/config/appservice.py
Comment (clokep) [reply synapse/config/appservice.py:150]: This comment is wrong. Strings are iterables, not lists.

File: synapse/appservice/__init__.py
Comment (clokep) [reply synapse/appservice/__init__.py]: Yeah, I removed that and left it as a `dict`, it simplified the diff considerably.

File: tests/appservice/test_appservice.py
Comment (clokep) [reply tests/appservice/test_appservice.py]: Yes, we should convert this to a `HomeserverTestCase` at some point, but definitely out of scope for adding types!

File: synapse/config/appservice.py
Comment (squahtx) [reply synapse/config/appservice.py:150]: Agreed, the comment is confused. Strings count as `Sequence`s, but they definitely aren't lists.

File: synapse/appservice/__init__.py
Comment (clokep) [reply synapse/appservice/__init__.py]: Done! I always forget `Pattern` can be used over bytes too.

File: synapse/config/appservice.py
Comment (clokep) [reply synapse/config/appservice.py:150]: (Do you want me to change anything about this?)

Comment (DMRobertson) [reply synapse/config/appservice.py:150]: Nope, I just wanted to flag up the old comment's wrongness.

**Review by DMRobertson — APPROVED** [review]
I'm happy with this, assuming CI completes successfully. Many thanks!

**Pull request merged**
